.github/workflows/pre-merge-e2e-checks.yml

```diff
@@ -0,0 +1,88 @@
+name: Pre-merge e2e checks
+
+permissions: {}
+
+env:
+  DOCKER_REGISTRY: nethermind.jfrog.io
+  REPO_DEV: nubia-oci-local-dev
+
+on:
+  merge_group:
+    branches: [main]
+
+concurrency:
+  group: shared_dev_environment
+  cancel-in-progress: false
+
+jobs:
+  build_docker_image:
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+    outputs:
+      DOCKER_IMAGE_TAG: ${{ steps.set_tag.outputs.DOCKER_IMAGE_TAG }}
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+
+      - name: Define image tag
+        id: set_tag
+        run: |
+          export DOCKER_IMAGE_TAG=$(git describe --tags)
+          echo "DOCKER_IMAGE_TAG=$DOCKER_IMAGE_TAG" >> $GITHUB_ENV
+          echo "DOCKER_IMAGE_TAG=$DOCKER_IMAGE_TAG" >> $GITHUB_OUTPUT
+
+      - name: Setup docker buildx
+        uses: docker/setup-buildx-action@f7ce87c1d6bead3e36075b2ce75da1f6cc28aaca
+
+      - name: Login to registry
+        run: |
+          docker login ${{ env.DOCKER_REGISTRY }} -u ${{ secrets.ARTIFACTORY_NUBIA_USERNAME }} -p ${{ secrets.ARTIFACTORY_NUBIA_TOKEN_DEVELOPER }}
+
+      - name: Build and Push
+        uses: docker/build-push-action@0adf9959216b96bec444f325f1e493d4aa344497
+        with:
+          context: .
+          platforms: "linux/amd64"
+          push: true
+          tags: |
+            ${{ env.DOCKER_REGISTRY }}/${{ env.REPO_DEV }}/juno:${{ env.DOCKER_IMAGE_TAG }}
+
+  validate_dev:
+    permissions:
+      id-token: write
+      contents: write
+    needs: [build_docker_image]
+    runs-on: ubuntu-latest
+    environment:
+      name: Development
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v4
+
+      - name: Verify Deployment Version (Dev)
+        run: |
+          bash .github/workflow-scripts/verify_deployment.sh ${{ secrets.DEV_SEPOLIA_URL }} ${{ needs.build_docker_image.outputs.DOCKER_IMAGE_TAG }}
+
+  dev-starknet-rs-tests:
+    needs: [validate_dev]
+    uses: ./.github/workflows/starknet-rs-tests.yml
+    secrets:
+      STARKNET_RPC: ${{ secrets.DEV_SEPOLIA_URL }}/v0_7
+
+  dev-starknet-js-tests:
+    needs: [validate_dev]
+    uses: ./.github/workflows/starknet-js-tests.yml
+    secrets:
+      TEST_RPC_URL: ${{ secrets.DEV_SEPOLIA_URL }}/v0_7
+      TEST_ACCOUNT_ADDRESS: ${{ secrets.TEST_ACCOUNT_ADDRESS }}
+      TEST_ACCOUNT_PRIVATE_KEY: ${{ secrets.TEST_ACCOUNT_PRIVATE_KEY }}
```

Comment:
Does it cause conflict if we run multiple instances of this test? I think we may consider adding job-level concurrency group for this.

Reply:
there is already a concurrency group for this workflow, so it won't run multiple instances of this test :)